File: javascript/problem023.js

var abundant = function(limit = 28123) {
	
	var divisorSums = (function(){
		
		var sieve = [-1];
		
		for (var i = 1; i < limit; i++)
			sieve.push(1);
				
		for (var a = 2; a < limit; a++)
			for (b = 2; b * a < limit; b++)
				sieve[a*b] += a;
			
		return sieve;
		
	});
	
	var abundant = (function() {
		
		var sums = divisorSums();
		var abundant = [];
		
		sums.forEach((sum, index) => { if (sum > index) abundant.push(index); });
		
		return abundant;
		
	})();
	
	var abundantSums = (function() { var arr = []; for (var i = 0; i < limit; i++) arr.push(false); return arr; })();
	
	for (var i = 0; i < limit; i++) {
		for (var j = 0; j <= i && abundant[i] + abundant[j] < limit ; j++) {
			abundantSums[abundant[i] + abundant[j]] = true;
		}
	}
	
	return abundantSums.reduce((sum, isAbundantSum, num) => isAbundantSum ? sum : sum + num, 0)
	
	
}

abundant();

File: python/problem023.py
# Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers

def d(num):
  divisors = [ d for d in range(2, int(num ** .5) + 1) if num % d == 0 ]
  return sum(set([1] + divisors + [ num // d for d in divisors ]))

limit = 28123

abun = [ num for num in range(12, limit) if d(num) > num ]
non = range(limit + 1)

for index, num1 in enumerate(abun):
  for num2 in abun[:index+1]:
    total = num1 + num2
    if total > limit: break
    non[total] = 0

print(sum(non))

